crates/cext/src/dag.rs

```diff
@@ -151,12 +157,553 @@ pub unsafe extern "C" fn qk_dag_num_qubits(dag: *const DAGCircuit) -> u32 {
+    }) as u32
+}
+
+/// @ingroup QkDag
+/// Gets the number of qubits of the specified operation node.
+///
+/// Panics if the node is not an operation.
+///
+/// @param dag A pointer to the DAG.
+/// @param node The operation node to get the number of qubits of.
+///
+/// @return The number of qubits of the operation.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_op_node_num_qubits(dag: *const DAGCircuit, node: u32) -> u32 {
+    // SAFETY: Per documentation, the pointer is to valid data.
+    let dag = unsafe { const_ptr_as_ref(dag) };
+    let instr = dag.dag()[NodeIndex::new(node as usize)].unwrap_operation();
+    instr.op.num_qubits()
+}
+
+/// @ingroup QkDag
+/// Gets the number of clbits of the specified operation node.
+///
+/// Panics if the node is not an operation.
+///
+/// @param dag A pointer to the DAG.
+/// @param node The operation node to get the number of clbits of.
+///
+/// @return The number of clbits of the operation.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_op_node_num_clbits(dag: *const DAGCircuit, node: u32) -> u32 {
+    // SAFETY: Per documentation, the pointer is to valid data.
+    let dag = unsafe { const_ptr_as_ref(dag) };
+    let instr = dag.dag()[NodeIndex::new(node as usize)].unwrap_operation();
+    instr.op.num_clbits()
+}
+
+/// @ingroup QkDag
+/// Gets the number of params of the specified operation node.
+///
+/// Panics if the node is not an operation.
+///
+/// @param dag A pointer to the DAG.
+/// @param node The operation node to get the number of params of.
+///
+/// @return The number of params of the operation.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_op_node_num_params(dag: *const DAGCircuit, node: u32) -> u32 {
+    // SAFETY: Per documentation, the pointer is to valid data.
+    let dag = unsafe { const_ptr_as_ref(dag) };
+    let instr = dag.dag()[NodeIndex::new(node as usize)].unwrap_operation();
+    instr.op.num_params()
+}
+
+/// @ingroup QkDag
+/// Retrieve the qubits of the specified operation node.
+///
+/// Panics if the node is not an operation.
+///
+/// @param dag A pointer to the DAG.
+/// @param node The operation node to get the qubits of.
+///
+/// @return A pointer to the qubits. Use ``qk_dag_op_node_num_qubits`` to determine the number of
+///     elements.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_op_node_qubits(dag: *const DAGCircuit, node: u32) -> *const u32 {
```

Comment:
For this method, since it returns a pointer to an array of `u32`s. Do we need to free these in the end or is it fine because they're already stored in the stack?

Reply:
Users don't need to free this because we're just "lending" a pointer to an allocation that remains backed by Rust.

That said, we should be very clear in the documentation that this point is borrowed by the DAG, and is only valid while the DAG's qubits are not modified.  (Strictly it's the interner that matters, but it's probably easier and safer against expansion to say "anything that modifies the qubits".)